### ISSUE TITLE
Error :  Function calls are not supported in decorators but 'QuillModule' was called.

### ISSUE THREAD

**Issue by balajikailasamintellect** [KillerCodeMonkey/ngx-quill#884]
I get this error only when I run a Production build, but development build works just fine. I am not sure if the problem is with AOT compile. My Angular version is 8.

I have the following entry in  my **app,modue.ts**

import { QuillModule } from 'ngx-quill';

imports: [
....
    QuillModule.forRoot(),
]

Entries in package.json
===============
"ngx-quill": "^9.1.1",
 "quill": "^1.3.7",
 "quill-mention": "^2.2.5",

Error Below
===============
![image](https://user-images.githubusercontent.com/62938168/81057992-26f0de80-8eeb-11ea-96fb-eecd90cde840.png)

really appreciate your help with this.
Thanks

**Comment by KillerCodeMonkey** [comment]
could you create a minimal demo repo for that please?

aot builds are working fine in general, checkout my demo repo, it is build with aot and production optimizations.

https://github.com/KillerCodeMonkey/ngx-quill-example
https://killercodemonkey.github.io/ngx-quill-example/

**Comment by KillerCodeMonkey** [comment]
ahh i see you should add Modules to `providers` and not `imports`

**Comment by langfors** [comment]
@KillerCodeMonkey I am getting the same error.  I'm sorry I don't understand your answer above.  Are you saying that QuillModule should be listed as a provider?  I tried this and still get this error.  Thanks

**Comment by KillerCodeMonkey** [comment]
i thought ths user added QuillModule.forRoot() to the `providers` array. but it seems you already added it as imports.

Please create a demo repo where i can see the and reproduce the error.

Maybe it has something to do with your tsconfig:
https://github.com/angular/angular/issues/23609

**Comment by KillerCodeMonkey** [comment]
But i will release a new version, hopefully it will be solved there.

**Comment by KillerCodeMonkey** [comment]
@langfors @balajikailasamintellect should not occure in latest version 10.0.1

**Comment by isipro** [comment]
Hello,

Same issue when building for production - being on the newest Quill version (ngx-quill@10.0.1)

ERROR in Error during template compile of 'AdminModule'
  Function calls are not supported in decorators but 'QuillModule' was called.
Unexpected value 'undefined' imported by the module 'AdminModule in C:/Projects/olemasport.com/WebSite/olemasport-com/src/app/admin/admin.module.ts'
Error during template compile of 'AdminModule'
  Function calls are not supported in decorators but 'QuillModule' was called.

> npm view ngx-quill
ngx-quill@10.0.1 | MIT | deps: none | versions: 198 


##ng version

Angular CLI: 8.3.3
Node: 10.16.0
OS: win32 x64
Angular: 8.2.5
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.803.3
@angular-devkit/build-angular     0.803.3
@angular-devkit/build-optimizer   0.803.3
@angular-devkit/build-webpack     0.803.3
@angular-devkit/core              8.3.3
@angular-devkit/schematics        8.3.3
@angular/cdk                      8.2.3
@angular/cli                      8.3.3
@angular/fire                     5.2.1
@angular/material                 8.2.1
@ngtools/webpack                  8.3.3
@schematics/angular               8.3.3
@schematics/update                0.803.3
rxjs                              6.4.0
typescript                        3.5.3
webpack                           4.39.2

Any advice?
Thanks in advance!

**Comment by KillerCodeMonkey** [comment]
How do you import ngx-quill?

And did you checked The example repo?

I would guess some Imports are wrong or you should check your tsconfig

Ivan Ivanov <notifications@github.com> schrieb am So., 10. Mai 2020, 18:55:

**Comment by KillerCodeMonkey** [comment]
Btw you could try to Upgrade typescript


Bengt Weiße <bengtler@googlemail.com> schrieb am So., 10. Mai 2020, 19:01:

**Comment by isipro** [comment]
@KillerCodeMonkey  
Thx for the quick answer!!!

Checked on the tsconfig.json
Indeed there is a difference when the 'fullTemlateTypeCheck is put to false.
The error is not smaller : )

ERROR in Error during template compile of 'AdminModule'
  Function calls are not supported in decorators but 'QuillModule' was called.

Might be related to this one 
https://github.com/angular/angular/issues/23609
where suggestion is that an upgrade to Angular 9 could work due to the Ivy compier.

so I put 
"enableIvy": true in "angularCompilerOptions" in tsconfig.json

ERROR in ./node_modules/ngx-quill/__ivy_ngcc__/fesm5/ngx-quill.js
Module build failed (from ./node_modules/@angular-devkit/build-optimizer/src/build-optimizer/webpack-loader.js):
TypeError: Cannot read property 'kind' of undefined

(I hope to not confuse anyone - last part is enabling Ivy on angular 8)

**Comment by KillerCodeMonkey** [comment]
So in general ngx-quill is build without ivy. And i would guess you have
Just to Upgrade typescript to 3.7.x


Ivan Ivanov <notifications@github.com> schrieb am So., 10. Mai 2020, 19:20:

**Comment by isipro** [comment]
tried with typescript 3.7.x and latest (3.8.x ) still the same

Found some info here
https://github.com/angular/angular/issues/23609
"""
From my experience I can tell that the issue is often produced in static forRoot(): ModuleWithProviders, sometimes depending on one line of code that becomes or becomes not "statically analyzable".
"""
NOTE: Works ok on a brand new angular 9 installation from today. 
So it should be something specific with my installation and versions (I will revert and install it again on the the angular 8 project and provide a demo repo )

**Comment by jpike88** [comment]
Im getting this error with latest everything, and I'm using ngcc to build, this is not related to my bundler.

**Comment by jpike88** [comment]
setting ```"fullTemplateTypeCheck": false``` has hidden the error...

**Comment by jpike88** [comment]
@KillerCodeMonkey I've done some more research. The reason that you haven't maybe noticed the issue is to do with the ```forRoot``` function in your module.


https://github.com/KillerCodeMonkey/ngx-quill/blob/c21f717a052cd5d2c88b3af827d5cd29e1e53000/projects/ngx-quill/src/lib/quill.module.ts#L22

Maybe it needs to be fully static, I think that it isn't because it's assigning a variable at line 26. If you could reopen this issue I'll investigate further

**Comment by jpike88** [comment]
work in progress... something is definitely wrong here

<img width="956" alt="Screen Shot 2020-05-13 at 7 24 46 PM" src="https://user-images.githubusercontent.com/9585787/81806685-694f9680-954f-11ea-963f-24ade39caa51.png">

**Comment by jpike88** [comment]
```config.modules = defaultModules;```

This fails static analysis, that's why its erroring. It needs to be moved

**Comment by jpike88** [comment]
@KillerCodeMonkey I think I fixed it, what do you think

**Comment by isipro** [comment]
As it turned out this to not be fun any longer 
I upgraded from Angular 8 to Angular 9 and it works for me ( previously tested plain new Ang9 - it also worked ) 

no issues like "Function calls are not supported in decorators ... " when doing ng build --prod

May be a good idea to go forward this way if understanding and applying all the provided fixes do not work for you either

Thank you!

**Comment by mcelicalderon** [comment]
@KillerCodeMonkey the fix wasn't released for v9.x right? I'm looking at the same problem there, any advice? Or v10.x is the only way to fix it?

**Comment by rohithgopip** [comment]
Issue still occurs in Angular 8.x

**Comment by KillerCodeMonkey** [comment]
v10 is the only way to fix it. because it is not possible to fix it that way for angular 8 :S

**Comment by rohithgopip** [comment]
V10 doesn't support Angular 8.x?

**Comment by KillerCodeMonkey** [comment]
The problem is I forgot to remove angular 8 as supported version in the latest releases. I should definitely do this in the next one.

**Comment by yuri-scarbaci-lenio** [comment]
You might want to update the "angular versions" support table on the readme,
We are using Angular 8 and since the readme stated the latest compatible version was < 10 we installed `^9.2.0` but that broke the building, which means Angular 8 surely is not compatible with version 9 either...

**Comment by yuri-scarbaci-lenio** [comment]
For whoever come to this and needs to support angular 8

Looks like `7.3.12` is the version to go

**Comment by mopcweb** [comment]
@yuri-scarbaci-lenio Thanks man!

**Comment by dkempkes** [comment]
> For whoever come to this and needs to support angular 8
> 
> Looks like `7.3.12` is the version to go

I can confirm that this is a valid solution. 

I will update tot a newer angular version when I find the time

**Comment by KieranMcKeefery** [comment]
> > For whoever come to this and needs to support angular 8
> > Looks like `7.3.12` is the version to go
> 
> I can confirm that this is a valid solution.
> 
> I will update tot a newer angular version when I find the time

Hey, how did you end up getting it working please? I'm struggling on this end - we're on 7.3 too, tried everything so far.

**Comment by dkempkes** [comment]
> > > For whoever come to this and needs to support angular 8
> > > Looks like `7.3.12` is the version to go
> > 
> > 
> > I can confirm that this is a valid solution.
> > I will update tot a newer angular version when I find the time
> 
> Hey, how did you end up getting it working please? I'm struggling on this end - we're on 7.3 too, tried everything so far.

Did you add the module QuillModule to your imports, if you're using lazy loaded modules? Aside from the QuillModule.forRoot(). If not, could you share some more info on your issue.

**Comment by Prasanth-FrontEnd** [comment]
> Had to update to Angular 9. Build working fine after that.

Even I have migrated to Angular 9 version but I'm getting the same error. What version of ngx-quill you are using in Angular 9 to fix this issue?
